pkg/cmd/experimental/syncgroups/sync-groups.go

```diff
@@ -0,0 +1,335 @@
+	// interpret user-provided configuration
+	clientConfig, err := ldaputil.CreateClientConfig(o.Config.LDAPClientConfig)
+	if err != nil {
+		return fmt.Errorf("could not determine LDAP client configuration: %v", err)
+	}
+
+	userAttributeDefiner := ldaputil.NewLDAPUserAttributeDefiner(o.Config.LDAPAttributeMapping)
+
+	// populate schema-independent syncer fields
+	syncer := LDAPGroupSyncer{
+		Host:           clientConfig.Host,
+		GroupClient:    o.GroupInterface,
+		UserNameMapper: NewUserNameMapper(userAttributeDefiner),
+	}
+
+	// We can determine the schema used in the LDAP server by determining which types of queries are necessary
+	// to recover groups and users from the server:
+	//   - An LDAP server with first-class groups and users where group membership is held in group attributes
+	//     will have a specific query template for groups and a specific query template from an attribute source
+	//     for users
+	// 	 - An LDAP server with first-class users only will have no query template for groups and a specific query
+	//     query template for the user entries that are members of a group
+	//   - An LDAP server with first-class groups and users where group membership is held in user attributes
+	//     will have a specific query from attribute template for groups and a specific query template for users
+	// Once the schema has been determined, the user-input sync configuration is interpreted and used to
+	// create a schema-specific LDAP interface which is instrumental in creating the final GroupSyncer.
+	switch schemaConfig := o.Config.LDAPConfig.Object.(type) {
+	case (*api.Schema1LDAPConfig):
+		// create syncer
+		// the schema-specific ldapInterface is built from the config
+		ldapInterface := schema1.NewLDAPInterface(clientConfig,
+			userAttributeDefiner,
+			schemaConfig.GroupQueryTemplate,
+			schemaConfig.UserQueryTemplate)
+
+		// The LDAPInterface knows how to extract group members
+		syncer.GroupMemberExtractor = &ldapInterface
+
+		// In order to build the GroupNameMapper, we need to know if the user defined a hard mapping
+		// or one based on LDAP group entry attributes
+		syncer.GroupNameMapper = getGroupNameMapper(o.Config.LDAPGroupUIDToOpenShiftNameMapping,
+			schemaConfig.GroupNameAttributes,
+			&ldapInterface)
+
+		// In order to build the groupLister, we need to know about the group sync scope and source:
+		syncer.GroupLister = getGroupLister(o.Scope,
+			o.Source,
+			o.WhitelistContents,
+			o.GroupInterface,
+			clientConfig.Host,
+			&ldapInterface)
+	default:
+		return fmt.Errorf("invalid schema-specific query template type: %v", schemaConfig)
+	}
+
+	// Now we run the Syncer and report any errors
+	syncErrors := syncer.Sync()
+	return kerrs.NewAggregate(syncErrors)
+}
+
+// getGroupNameMapper returns an LDAPGroupNameMapper either by using a user-provided mapping or
+// by creating a mapper that uses an algorithmic mapping
+func getGroupNameMapper(userDefinedMapping map[string]string,
```

Comment:
I don't understand this comment. What is a `defaultGroupNameMapper` ? When should it be chosen (to be regurgitated by the method?) instead of constructing new mappers from the config?